### PR TITLE
Add autoinstall flatpaks to update USBs

### PR DESCRIPTION
**Pull request by mwleeds**
When an OS update has flatpak(s) set to be updated or installed along
with it (via eos-updater-flatpak-installer), the flatpaks are required
for the update for succeed. So this commit adds the flatpaks to a USB
being created with eos-updater-prepare-volume so that USB updates work
even when the update in question has flatpaks to be autoinstalled. This
is accomplished by adding a "print" mode to
eos-updater-flatpak-installer and parsing the output from that in
eos-updater-prepare-volume. This way the logic that parses the
configuration files can stay in one place. Since we can't make any
assumptions about the OS version on the client receiving the USB update,
every flatpak configured to be automatically updated or installed on
OS update must be included.

https://phabricator.endlessm.com/T21756

### CODE REVIEW

**Comment by mwleeds** [comment]
I'm not super happy with the way this is written, especially since it would break if a prefix other than `/usr` is used, but I'm submitting the PR anyway as a starting point for discussion.

File: eos-updater-flatpak-installer/main.c [hunk 1]
```diff
@@ -249,6 +249,37 @@ main (int    argc,
...
+          if (new_flatpak_ref_actions_to_apply == NULL)
+            return fail (EXIT_FAILED,
+                         "Could not get information on which flatpak ref actions to apply: %s",
+                         error->message);
+
+          formatted_flatpak_ref_actions_to_apply =
+            euu_format_all_flatpak_ref_actions ("All flatpak ref actions that are not yet applied",
+                                                new_flatpak_ref_actions_to_apply);
+          g_print ("%s\n", formatted_flatpak_ref_actions_to_apply);
+
```
Comment (smspillaz) [inline eos-updater-flatpak-installer/main.c]: Same as above, can you split this out into another commit?

Comment (smspillaz) [reply eos-updater-flatpak-installer/main.c]: It strikes me that this probably repeats a lot of the existing debug messages in the tool. I wonder if it would make more sense to add a `--dry-run`option.

File: eos-updater-prepare-volume/eos-updater-prepare-volume [hunk 2]
```diff
@@ -173,6 +174,17 @@ class VolumePreparer:
...
+    def _add_autoinstall_flatpaks(self, flatpak_collection_refs):
+        all_actions = self.__run(['/usr/libexec/eos-updater-flatpak-installer', '--mode=print'])
+        for line in all_actions.split('\n'):
+            pattern = '^ *- (install|update) \(collection-id: (?P<cid>.*?)\|.*?\):(?P<ref>.*?)$'
+            match = re.search(pattern, line)
+            if match is not None:
+                ref_tuple = (match.group('cid'), match.group('ref'))
+                if self.debug:
+                    print('Adding a Flatpak autoinstall app: ' + str(ref_tuple))
+                flatpak_collection_refs.append(ref_tuple)
```
Comment (smspillaz) [inline eos-updater-prepare-volume/eos-updater-prepare-volume]: Oh, I see how this is being used now.

Instead of parsing the output of `--print` (which was adapted from the existing debug messages), perhaps it might be better to have `eos-updater-flatpak-installer` dump JSON and use that instead?

File: eos-updater-prepare-volume/eos-updater-prepare-volume [hunk 3]
```diff
@@ -173,6 +174,17 @@ class VolumePreparer:
 
         return runtimes
 
+    def _add_autoinstall_flatpaks(self, flatpak_collection_refs):
+        all_actions = self.__run(['/usr/libexec/eos-updater-flatpak-installer', '--mode=print'])
```
Comment (smspillaz) [inline eos-updater-prepare-volume/eos-updater-prepare-volume]: `libexec` is probably not correct here - `$(libexecdir)` can change depending on whether or not a debian package is being built (for instance, all the libexec stuff in debian packages gets installed into /usr/lib/arch-triple

You probably want to either preprocess this file or have a `config.py.in` where you can substitute the correct `$(libexecdir)` from autotools.

File: libeos-updater-util/flatpak.c [hunk 4]
```diff
@@ -1578,6 +1578,9 @@ format_remote_ref_action_type (EuuFlatpakRemoteRefActionType action_type)
   return enum_value->value_nick;
 }
 
+/* NOTE: The output produced by this function is parsed by
+ * eos-updater-prepare-volume, so be mindful of that if you make any changes!
+ */
```
Comment (smspillaz) [inline libeos-updater-util/flatpak.c]: It might be better to dump JSON in this case :)

File: eos-updater-prepare-volume/eos-updater-prepare-volume (hunk 2 above)
Comment (smspillaz) [reply eos-updater-prepare-volume/eos-updater-prepare-volume]: I was also thinking that perhaps `--print` might not belong in `eos-updater-flatpak-installer`. The good news is that `eufi_determine_flatpak_ref_actions_to_check` and friends are already in a separate library, so you could create a new binary and link directly to that. Not sure if we want yet another binary though. @pwithnall any thoughts on that?

File: eos-updater-prepare-volume/eos-updater-prepare-volume (hunk 3 above)
Comment (mwleeds) [reply eos-updater-prepare-volume/eos-updater-prepare-volume]: That's a good thought. I thought about preprocessing but I didn't think about doing it in a separate Python file. That will be neater.

File: libeos-updater-util/flatpak.c (hunk 4 above)
Comment (mwleeds) [reply libeos-updater-util/flatpak.c]: On the one hand that's added complexity and one more thing to maintain, but I suppose it's worth it to have something machine readable and a bit more robust.

File: eos-updater-prepare-volume/eos-updater-prepare-volume (hunk 2 above)
Comment (smspillaz) [reply eos-updater-prepare-volume/eos-updater-prepare-volume]: FWIW, I agree with @pwithnall , though I don't know if libeos-updater-flatpak-installer is introspected yet. Shouldn't be too bad to introspect it, though it does use GPtrArray quite heavily and not sure if the element types would need to be boxed.

File: Makefile.am [hunk 5]
```diff
@@ -295,6 +295,47 @@ libeos_updater_flatpak_installer_libeos_updater_flatpak_installer_@EUFI_API_VERS
...
 libexec_PROGRAMS += eos-updater-flatpak-installer/eos-updater-flatpak-installer
 
+###############################################################################
+# Introspection
+-include $(INTROSPECTION_MAKEFILE)
+INTROSPECTION_GIRS =
+INTROSPECTION_SCANNER_ARGS =
+INTROSPECTION_COMPILER_ARGS =
+AM_DISTCHECK_CONFIGURE_FLAGS += --enable-introspection
+EXTRA_DIST += m4/introspection.m4
```
Comment (pwithnall) [inline Makefile.am]: You need to add `m4/introspection.m4` to git.

File: Makefile.am [hunk 6]
```diff
@@ -295,6 +295,47 @@ libeos_updater_flatpak_installer_libeos_updater_flatpak_installer_@EUFI_API_VERS
...
+
+if HAVE_INTROSPECTION
+libeos-updater-flatpak-installer/Libeufi-@EUFI_API_VERSION@.gir: libeos-updater-flatpak-installer/libeos-updater-flatpak-installer-@EUFI_API_VERSION@.la
```
Comment (pwithnall) [inline Makefile.am]: By convention, `.gir` files don’t have a `Lib` prefix — so this should probably be `EosUpdaterFlatpakInstaller-@EUFI_API_VERSION@.gir`.

Comment (pwithnall) [reply Makefile.am]: For introspection to work, the library needs to be linkable against, which means you need to change `noinst_LTLIBRARIES += libeos-updater-flatpak-installer/libeos-updater-flatpak-installer-@EUFI_API_VERSION@.la` above to `lib_LTLIBRARIES += libeos-updater-flatpak-installer/libeos-updater-flatpak-installer-@EUFI_API_VERSION@.la`.

File: Makefile.am [hunk 7]
```diff
@@ -295,6 +295,47 @@ libeos_updater_flatpak_installer_libeos_updater_flatpak_installer_@EUFI_API_VERS
...
+libeos_updater_flatpak_installer_Libeufi_@EUFI_API_VERSION@_gir_INCLUDES = GObject-2.0 Gio-2.0
+libeos_updater_flatpak_installer_Libeufi_@EUFI_API_VERSION@_gir_CFLAGS = \
+	$(libeos_updater_flatpak_installer_libeos_updater_flatpak_installer_@EUFI_API_VERSION@_la_CFLAGS) \
```
Comment (pwithnall) [inline Makefile.am]: I think this should be:
```
$(STD_CFLAGS) \
$(EOS_UPDATER_FLATPAK_INSTALLER_CFLAGS) \
$(NULL)
```
rather than using the full list of `CFLAGS` from the library, otherwise you’ll end up with warnings and code coverage enabled for the GIR scanner, which will probably break things. GIR generates a C file which it compiles and runs to extract the runtime `GType` information from a library. Since it’s generated code, compiling it with warnings enabled is generally a bad idea.

File: Makefile.am [hunk 8]
```diff
@@ -295,6 +295,47 @@ libeos_updater_flatpak_installer_libeos_updater_flatpak_installer_@EUFI_API_VERS
...
+	$(NULL)
+libeos_updater_flatpak_installer_Libeufi_@EUFI_API_VERSION@_gir_FILES =
+	libeos-updater-flatpak-installer/determine-flatpak-actions.c \
+	libeos-updater-flatpak-installer/installer.h \
+	libeos-updater-flatpak-installer/perform-flatpak-actions.c \
```
Comment (pwithnall) [inline Makefile.am]: This list could just be `$(libeos_updater_flatpak_installer_libeos_updater_flatpak_installer_@EUFI_API_VERSION@_la_SOURCES)`.

File: Makefile.am [hunk 9]
```diff
@@ -295,6 +295,47 @@ libeos_updater_flatpak_installer_libeos_updater_flatpak_installer_@EUFI_API_VERS
...
+	$(NULL)
+libeos_updater_flatpak_installer_Libeufi_@EUFI_API_VERSION@_gir_LIBS = libeos-updater-flatpak-installer/libeos-updater-flatpak-installer-@EUFI_API_VERSION@.la
+libeos_updater_flatpak_installer_Libeufi_@EUFI_API_VERSION@_gir_NAMESPACE = Libeufi
```
Comment (pwithnall) [inline Makefile.am]: `EosUpdaterFlatpakInstaller`

File: Makefile.am [hunk 10]
```diff
@@ -295,6 +295,47 @@ libeos_updater_flatpak_installer_libeos_updater_flatpak_installer_@EUFI_API_VERS
...
+libeos_updater_flatpak_installer_Libeufi_@EUFI_API_VERSION@_gir_SCANNERFLAGS = \
+	--nsversion=@EUFI_API_VERSION@ \
+	--warn-all \
```
Comment (pwithnall) [inline Makefile.am]: Drop `--warn-all` if you’re using `$(WARN_SCANNERFLAGS)`.

File: Makefile.am [hunk 11]
```diff
@@ -295,6 +295,47 @@ libeos_updater_flatpak_installer_libeos_updater_flatpak_installer_@EUFI_API_VERS
...
+	--c-include="libeos-updater-flatpak-installer/installer.h" \
```
Comment (pwithnall) [inline Makefile.am]: Also add `--identifier-prefix=Eufi --symbol-prefix=eufi`.

File: Makefile.am (hunk 6 above)
Comment (pwithnall) [reply Makefile.am]: > By convention, …

Ah, it seems you already did that in the follow up commit. :+1:

**Comment by mwleeds** [comment]
This is still WIP, but I put a status update on the ticket.

**Comment by mwleeds** [comment]
This is ready for re-review now.

**Comment by mwleeds** [comment]
There is a critical warning at runtime:
`** (process:6018): CRITICAL **: Unsupported type array`

But I'm not sure it's anything to lose sleep over. The array is clearly usable by Python.

**Comment by pwithnall** [comment]
> There is a critical warning at runtime:
> `** (process:6018): CRITICAL **: Unsupported type array`
> 
> But I'm not sure it's anything to lose sleep over. The array is clearly usable by Python.

It’s definitely something I’d want fixed before merging. Run under `gdb` with `G_DEBUG=fatal-warnings` in the environment, and you should get a backtrace from where the critical’s being emitted, which should give you some clues about what’s causing it. Poke me in Slack if you need a second opinion.

wrt the Jenkins failure:

```
Couldn't find include 'EosUpdaterUtil-0.gir' (search path: '['gir-1.0', '/usr/share/gir-1.0', '/usr/share/gir-1.0', '/usr/share/gir-1.0']')
libtool: link: ( cd "test-common/.libs" && rm -f "libeos-updater-test-common.la" && ln -s "../libeos-updater-test-common.la" "libeos-updater-test-common.la" )
/usr/share/gobject-introspection-1.0/Makefile.introspection:159: recipe for target 'libeos-updater-flatpak-installer/EosUpdaterFlatpakInstaller-0.gir' failed
```

That can be fixed by adding `--add-include-path=$(top_builddir)/libeos-updater-util` to the relevant `SCANNERFLAGS`.

File: eos-updater-prepare-volume/eos-updater-prepare-volume [hunk 12]
```diff
@@ -173,6 +176,13 @@ class VolumePreparer:
 
         return runtimes
 
+    def _add_autoinstall_flatpaks(self, flatpak_collection_refs):
+        applied_actions = EosUpdaterFlatpakInstaller.determine_flatpak_ref_actions_to_check(None)
```
Comment (pwithnall) [inline eos-updater-prepare-volume/eos-updater-prepare-volume]: Not entirely sure this is the right function to call. `determine_flatpak_ref_actions_to_check()` will load the autoinstall lists, parse them, and then *filter out* any entries whose serial number is lower than the latest applied serial number for this machine. I think you probably want to take part of the implementation for `determine_flatpak_ref_actions_to_check()` and call that instead. i.e. Call `flatpak_ref_actions_and_progresses()` and not `euu_filter_for_existing_flatpak_ref_actions()`. Since `flatpak_ref_actions_and_progresses()` is not public, it’s probably best to just copy part of its implementation (the bit relating to `out_actions`) instead. i.e. Call `euu_flatpak_ref_actions_from_paths()`.

Also, the above paragraph aside, did you consider calling `euu_flatten_flatpak_ref_actions_table()` on the `applied_actions` and reject the idea on the basis that the computer the updates are being applied to might have a different set of autoinstall list overrides, and hence it would be better to copy apps onto the USB stick even if their autoinstall list entries were overridden (squashed) on this machine?

File: eos-updater-prepare-volume/eos-updater-prepare-volume [hunk 13]
```diff
@@ -173,6 +176,13 @@ class VolumePreparer:
...
+        for filename in applied_actions:
+            for action in applied_actions[filename]:
```
Comment (pwithnall) [inline eos-updater-prepare-volume/eos-updater-prepare-volume]: I think you can omit a hash operation on the dict key here:
```
for filename, actions in applied_actions.items():
    for action in actions:
        …
```

File: eos-updater-prepare-volume/eos-updater-prepare-volume [hunk 14]
```diff
@@ -202,6 +212,9 @@ class VolumePreparer:
             return self.__fail(self.EXIT_INVALID_ARGUMENTS,
                                'Invalid flatpak collection–refs: %s' % refs_list)
 
+        # Add the flatpaks that will be installed by eos-updater-flatpak-installer
+        self._add_autoinstall_flatpaks(flatpak_collection_refs)
```
Comment (pwithnall) [inline eos-updater-prepare-volume/eos-updater-prepare-volume]: Need to handle exceptions from this call, so that the error message can be formatted nicely and a suitable exit status (`self.EXIT_FAILED`) returned.

File: eos-updater-prepare-volume/eos-updater-prepare-volume (hunk 13 above)
Comment (pwithnall) [reply eos-updater-prepare-volume/eos-updater-prepare-volume]: But this code will change once you’ve called `euu_flatten_flatpak_ref_actions_table()` on the `applied_actions` anyway.

File: libeos-updater-util/flatpak.c [hunk 15]
```diff
@@ -121,6 +121,18 @@ euu_flatpak_remote_ref_action_unref (EuuFlatpakRemoteRefAction *action)
   g_slice_free (EuuFlatpakRemoteRefAction, action);
 }
 
+char *
```
Comment (pwithnall) [inline libeos-updater-util/flatpak.c]: Nitpick: The eos-updater coding style uses `gchar` rather than `char`.

Comment (pwithnall) [reply libeos-updater-util/flatpak.c]: Please add gtk-doc comments for these new functions, too, since this is now essentially a public library, so we should be working towards full documentation coverage.

File: libeos-updater-util/flatpak.c [hunk 16]
```diff
@@ -121,6 +121,18 @@ euu_flatpak_remote_ref_action_unref (EuuFlatpakRemoteRefAction *action)
...
+euu_flatpak_remote_ref_action_get_collection_id (EuuFlatpakRemoteRefAction *action)
+{
+    return g_strdup (action->ref->collection_id);
+}
+
+char *
+euu_flatpak_remote_ref_action_get_ref (EuuFlatpakRemoteRefAction *action)
```
Comment (pwithnall) [inline libeos-updater-util/flatpak.c]: I’d call this `euu_flatpak_remote_ref_action_format_ref()`, since it’s returning the string formatted version of the ref, not the `FlatpakRef*` object.

Same comments from above about `gchar` and documentation apply here. Please make sure to mention that the formatted ref is generated by `flatpak_ref_format_ref()` and hence can be parsed by `flatpak_ref_parse()`.

**Comment by pwithnall** [comment]
Also, there are quite a few remaining warnings from `g-ir-scanner` in [the log](https://ci.endlessm-sf.com/job/obs-eos-updater-pr/222/console). Can you please file a follow-up Phabricator task about fixing them, if you haven’t already? Thanks.

**Comment by mwleeds** [comment]
> Also, there are quite a few remaining warnings from g-ir-scanner in the log. Can you please file a follow-up Phabricator task about fixing them, if you haven’t already? Thanks.

https://phabricator.endlessm.com/T22636

File: test-common/utils.c [hunk 17]
```diff
@@ -2932,12 +2932,17 @@ eos_test_client_prepare_volume (EosTestClient *client,
                                                                                "eos-updater-prepare-volume",
                                                                                "eos-updater-prepare-volume",
                                                                                NULL);
+  g_autofree gchar *libeos_updater_util_path = g_test_build_filename (G_TEST_DIST,
```
Comment (pwithnall) [inline test-common/utils.c]: Shouldn’t this be s/`G_TEST_DIST`/`G_TEST_BUILT`/, since the typelib is generated at build time, rather than shipped in git?

**Comment by pwithnall** [comment]
I’m going to look into the [runtime warning](https://github.com/endlessm/eos-updater/pull/188#issuecomment-389413696) and will do some work on this branch to fix up the comments while I’m at it, so please don’t do any more work on this yourself for the moment!

File: libeos-updater-util/flatpak.c (hunk 15 above)
Comment (pwithnall) [reply libeos-updater-util/flatpak.c]: Also, the indentation is wrong inside these functions.

**Comment by pwithnall** [comment]
OK, I’ve got some fixups [here](https://github.com/mwleeds/eos-updater/pull/1) which could be merged into this PR. I’d then like @smspillaz to take a look, [specifically at this bit](https://github.com/endlessm/eos-updater/pull/188#discussion_r188616028) (if he’s around and able).

If this PR gets accepted, before it’s merged, there are a few things to do:
 * Fix the [typo in a commit message](https://github.com/endlessm/eos-updater/pull/188#pullrequestreview-120624815)
 * Squash together [these two commits](https://github.com/endlessm/eos-updater/pull/188#pullrequestreview-120636020)
 * Ensure “Remove unused function parameter” is ordered before “Fix some g-ir-scanner warnings”
 * Rebase and squash in all the fixups (which will be non-trivial; it should work if they’re squashed in one-by-one, but squashing them in all at once causes conflicts because git doesn’t do it in order)

---

I looked at the critical warning message. It’s easily reproducible by running `python3` and then entering:
```
from gi.repository import EosUpdaterFlatpakInstaller, EosUpdaterUtil
EosUpdaterFlatpakInstaller.determine_flatpak_ref_actions_to_check(['/opt/endless/source/eos-application-tools/data'])
```

It gives the following backtrace:
```
** (process:31455): CRITICAL **: Unsupported type array

Program received signal SIGTRAP, Trace/breakpoint trap.
_g_log_abort (breakpoint=1) at /opt/endless/source/glib/glib/gmessages.c:554
554	}
(gdb) bt
#0  _g_log_abort (breakpoint=1) at /opt/endless/source/glib/glib/gmessages.c:554
#1  0x00007fffeebe9b3a in g_logv (log_domain=0x0, log_level=G_LOG_LEVEL_CRITICAL, 
    format=0x7fffef37fec8 "Unsupported type %s", args=0x7fffffffcbb8)
    at /opt/endless/source/glib/glib/gmessages.c:1362
#2  0x00007fffeebe9c2b in g_log (log_domain=0x0, log_level=G_LOG_LEVEL_CRITICAL, 
    format=0x7fffef37fec8 "Unsupported type %s") at /opt/endless/source/glib/glib/gmessages.c:1403
#3  0x00007fffef37ae47 in _pygi_marshal_to_py_ghash ()
   from /usr/lib64/python3.6/site-packages/gi/_gi.cpython-36m-x86_64-linux-gnu.so
#4  0x00007fffef371e4e in pygi_invoke_c_callable ()
   from /usr/lib64/python3.6/site-packages/gi/_gi.cpython-36m-x86_64-linux-gnu.so
#5  0x00007fffef3739b8 in pygi_function_cache_invoke ()
   from /usr/lib64/python3.6/site-packages/gi/_gi.cpython-36m-x86_64-linux-gnu.so
#6  0x00007ffff79980f4 in _PyObject_FastCallDict () from /lib64/libpython3.6m.so.1.0
#7  0x00007ffff7a06d8d in call_function () from /lib64/libpython3.6m.so.1.0
#8  0x00007ffff7a30e9a in _PyEval_EvalFrameDefault () from /lib64/libpython3.6m.so.1.0
#9  0x00007ffff7a0fb38 in PyEval_EvalCodeEx () from /lib64/libpython3.6m.so.1.0
#10 0x00007ffff7a108bb in PyEval_EvalCode () from /lib64/libpython3.6m.so.1.0
#11 0x00007ffff7a9c692 in run_mod () from /lib64/libpython3.6m.so.1.0
#12 0x00007ffff7967101 in PyRun_InteractiveOneObjectEx () from /lib64/libpython3.6m.so.1.0
#13 0x00007ffff7969bf6 in PyRun_InteractiveLoopFlags () from /lib64/libpython3.6m.so.1.0
#14 0x00007ffff7a9efc5 in PyRun_AnyFileExFlags () from /lib64/libpython3.6m.so.1.0
#15 0x00007ffff796ac61 in Py_Main.cold.2848 () from /lib64/libpython3.6m.so.1.0
#16 0x0000555555554e22 in main ()
```

Looking at that, and at the pygobject code, it seems that pygobject just doesn’t like having a `GPtrArray` as the value type of a `GHashTable`. As Matthew says, it does function perfectly well. I can’t see why the implementation of `GI_TYPE_TAG_ARRAY` should differ from the implementation of `GI_TYPE_TAG_INTERFACE` (GObject instances); no refcounting or checks are done in the latter case, so we shouldn’t need to do any for arrays. I think this is a bug in pygobject, and it should just not be warning about arrays.

⇒ I’ve filed https://gitlab.gnome.org/GNOME/pygobject/merge_requests/73 about it

File: libeos-updater-flatpak-installer/perform-flatpak-actions.c [hunk 18]
```diff
@@ -440,6 +456,20 @@ check_if_flatpak_is_installed (FlatpakInstallation        *installation,
   return TRUE;
 }
 
+/**
+ * eufi_check_ref_actions_applied:
+ * @installation: a #FlatpakInstallation
+ * @pending_flatpak_deployments_state_path: unused
```
[the remark this answers is not in the document]
Comment (smspillaz) [reply libeos-updater-flatpak-installer/perform-flatpak-actions.c]: It could be a bug. If that parameter was there for anything, it would be there to limit the checked actions to include the ones up to and including the serials specified in that path. Perpahps better to just remove it though

File: eos-updater-prepare-volume/eos-updater-prepare-volume (hunk 12 above)
Comment (smspillaz) [reply eos-updater-prepare-volume/eos-updater-prepare-volume]: In light of what @pwithnall here says about the overrides being potentially different on the target, the best bet here is to read in all the actions that might be necessary on the ostree and squash them from there.

I'm not sure how much can be done to handle the overrides situation - the source cant know what ref actions will be needed on the target since /var and /etc are not really managed by ostree

File: libeos-updater-util/flatpak.c [hunk 19]
```diff
@@ -1063,6 +1067,13 @@ euu_flatpak_ref_actions_from_data (const gchar   *data,
   return g_steal_pointer (&actions);
 }
 
+/**
+ * euu_flatpak_remote_ref_actions_file_new:
+ * @remote_ref_actions: (element-type EuuFlatpakRemoteRefAction): A #GPtrArray
+ * @priority: the priority of the file
```
[the remark this answers is not in the document]
Comment (smspillaz) [reply libeos-updater-util/flatpak.c]: Higher numbers take priority, See `test_update_deploy_flatpaks_on_reboot_override_ostree`. So the docs should be adjusted

[the remark this answers is not in the document]
Comment (pwithnall) [reply libeos-updater-util/flatpak.c]: The documentation (and implementation) for `euu_flatpak_ref_actions_append_from_directory()` have lower numeric values as more important. This is the priority which comes from the directory’s position in the `directories_to_search` in `euu_flatpak_ref_actions_from_paths()`; not the priority which comes from the file’s lexicographical position within its directory.

**Comment by mwleeds** [comment]
I'll leave this for @pwithnall to merge after looking at my changes to "libeuu: Ignore non-JSON files when loading autoinstall files from a dir"
